File: swift/Sources/FundamentalCore/Engine/ParticleShape.swift
import Foundation // cos/sin — from Darwin on Apple, Glibc on Linux (the simd shim doesn't provide them)
#if canImport(simd)
import simd
#endif

/// How a particle is drawn — a vector shape the renderer stamps at each particle's position, scaled by
/// its (physics-driven) size and heat. The default `.dot` keeps the fast batched-circle path; every
/// other shape is a unit polygon (origin-centred, radius ≤ 1) the renderer transforms per particle.
///
/// Renderer-agnostic data: the core only carries the vertices — each plane's renderer (CoreGraphics,
/// Metal, canvas, WebGL) draws them. Because the renderer scales the shape by each particle's `size`
/// and `heat`, the appearance still rides the physics: a hot particle is a bigger, brighter star.
///
/// ```swift
/// FieldView(options: .init(render: .dots, particleShape: .star(points: 5)))
/// // or a custom vector:  .custom([SIMD2(0, -1), SIMD2(1, 1), SIMD2(-1, 1)])  // a triangle
/// ```
public struct ParticleShape: Equatable, Sendable {
    /// Unit vertices in `[-1, 1]` (origin-centred), filled in order — or `nil` for the built-in circle
    /// (`.dot`), the fast batched path renderers keep accelerated.
    public let vertices: [SIMD2<Float>]?

    public init(vertices: [SIMD2<Float>]?) { self.vertices = vertices }

    /// The default soft circle — the fast batched path (Metal-accelerated where available).
    public static let dot = ParticleShape(vertices: nil)

    /// An N-pointed star. `innerRatio` ∈ (0, 1] sets the waist depth (0.5 = a classic five-point star).
    public static func star(points: Int = 5, innerRatio: Float = 0.5) -> ParticleShape {
        let n = max(2, points)
        let inner = min(max(innerRatio, 0.01), 1)
        var v = [SIMD2<Float>]()
        v.reserveCapacity(n * 2)
        for i in 0..<(n * 2) {
            let r = i.isMultiple(of: 2) ? Float(1) : inner
            let a = Float(i) * .pi / Float(n) - .pi / 2 // first point straight up
            v.append(SIMD2(cos(a) * r, sin(a) * r))
        }
        return ParticleShape(vertices: v)
    }

    /// A regular polygon — triangle (3), square (4), hexagon (6)… `rotation` in degrees.
    public static func polygon(sides: Int, rotation: Float = 0) -> ParticleShape {
        let n = max(3, sides)
        let rot = rotation * .pi / 180
        var v = [SIMD2<Float>]()
        v.reserveCapacity(n)
        for i in 0..<n {
            let a = Float(i) * 2 * .pi / Float(n) - .pi / 2 + rot
            v.append(SIMD2(cos(a), sin(a)))
        }
        return ParticleShape(vertices: v)
    }

    /// Any custom vector shape — your own unit vertices in `[-1, 1]` (origin-centred), filled in order.
    public static func custom(_ vertices: [SIMD2<Float>]) -> ParticleShape {
        ParticleShape(vertices: vertices)
    }

    /// Whether this is the fast built-in circle (no polygon path).
    public var isDot: Bool { vertices == nil }
}

File: android/fundamental-core/src/main/kotlin/com/fundamental/core/engine/ParticleShape.kt
package com.fundamental.core.engine

import com.fundamental.core.math.Vec3
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

// ParticleShape (particle-shape.ts) — the Kotlin port of
// swift/Sources/FundamentalCore/Engine/ParticleShape.swift. How a particle is drawn: a unit vector
// shape the renderer stamps at each particle's position, scaled by its (physics-driven) size + heat,
// so a hot particle is a bigger, brighter star. Renderer-agnostic data — the core only carries the
// vertices; each host's renderer transforms them. `dot` keeps the fast batched-circle path.
//
// Vertices are origin-centred in [-1, 1]. 2D shapes carried as `Vec3` with z = 0 (the engine is
// 3D-native; particle stamps are planar).

data class ParticleShape(val vertices: List<Vec3>?) {
    /** Whether this is the fast built-in circle (no polygon path). */
    val isDot: Boolean get() = vertices == null

    companion object {
        /** The default soft circle — the fast batched path. */
        val DOT = ParticleShape(null)

        /** An N-pointed star. `innerRatio` ∈ (0,1] sets the waist depth (0.5 = a classic 5-point star). */
        fun star(points: Int = 5, innerRatio: Float = 0.5f): ParticleShape {
            val n = max(2, points)
            val inner = min(max(innerRatio, 0.01f), 1f)
            val v = ArrayList<Vec3>(n * 2)
            for (i in 0 until n * 2) {
                val r = if (i % 2 == 0) 1f else inner
                val a = i.toFloat() * PI.toFloat() / n - PI.toFloat() / 2f // first point straight up
                v.add(Vec3(cos(a) * r, sin(a) * r, 0f))
            }
            return ParticleShape(v)
        }

        /** A regular polygon — triangle (3), square (4), hexagon (6)… `rotation` in degrees. */
        fun polygon(sides: Int, rotation: Float = 0f): ParticleShape {
            val n = max(3, sides)
            val rot = rotation * PI.toFloat() / 180f
            val v = ArrayList<Vec3>(n)
            for (i in 0 until n) {
                val a = i.toFloat() * 2f * PI.toFloat() / n - PI.toFloat() / 2f + rot
                v.add(Vec3(cos(a), sin(a), 0f))
            }
            return ParticleShape(v)
        }

        /** Any custom vector shape — your own unit vertices in [-1,1] (origin-centred), filled in order. */
        fun custom(vertices: List<Vec3>): ParticleShape = ParticleShape(vertices)
    }
}
